pkg/ttl/cache/table.go

```diff
@@ -189,24 +189,64 @@ func (t *PhysicalTable) ValidateKeyPrefix(key []types.Datum) error {
 func (t *PhysicalTable) EvalExpireTime(ctx context.Context, se session.Session,
 	now time.Time) (expire time.Time, err error) {
 	tz := se.GetSessionVars().Location()
+	now = now.In(tz)
 
 	expireExpr := t.TTLInfo.IntervalExprStr
 	unit := ast.TimeUnitType(t.TTLInfo.IntervalTimeUnit)
 
+	if now.Unix() < 0 {
+		return time.Time{}, errors.Errorf(
+			"current time must be after the UTC time January 1, 1970, 00:00:00. Received time: %s",
+			now,
+		)
+	}
+
 	var rows []chunk.Row
 	rows, err = se.ExecuteSQL(
 		ctx,
-		// FROM_UNIXTIME does not support negative value, so we use `FROM_UNIXTIME(0) + INTERVAL <current_ts>`
-		// to present current time
-		fmt.Sprintf("SELECT FROM_UNIXTIME(0) + INTERVAL %d SECOND - INTERVAL %s %s", now.Unix(), expireExpr, unit.String()),
+		fmt.Sprintf("SELECT FROM_UNIXTIME(%d) - INTERVAL %s %s", now.Unix(), expireExpr, unit.String()),
 	)
 
 	if err != nil {
 		return
 	}
 
 	tm := rows[0].GetTime(0)
-	return tm.CoreTime().GoTime(tz)
+	formatted, err := tm.DateFormat("%Y-%m-%d %H:%i:%s.%f")
+	if err != nil {
+		return time.Time{}, err
+	}
+
+	expire, err = time.ParseInLocation("2006-01-02 15:04:05.000000", formatted, tz)
+	if err != nil {
+		return time.Time{}, err
+	}
+
+	if expire.Hour() != tm.Hour() {
+		// For the daylight-saving time, the hour may be different, and it means the tm we get is not a "valid" date.
+		// For example, if we computed an expired `types.Time` with value "2024-03-10 02:30:00",
+		// this time does not "exist" in timezone `America/Los_Angeles` because after "2024-03-10 02:00:00",
+		// the time will leap to "2024-03-10 03:00:00" directly.
+		// If then we parse "2024-03-10 02:30:00" with `time.ParseInLocation` we will get a Go time "2024-03-10 01:30:00".
+		// So, return the parsed time directly.
+		return expire, nil
+	}
```

Comment:
There is also a weird time zone with 30 min offset for Day Light Saving Time: https://en.wikipedia.org/wiki/List_of_tz_database_time_zones#LORD_HOWE

For all the different time zones:
https://en.wikipedia.org/wiki/List_of_tz_database_time_zones